### ISSUE TITLE
PUT photometry error "UniqueViolation"

### ISSUE THREAD

**Issue by AlexGKim** [skyportal/skyportal#1899]
I am trying to use PUT photometry to update existing photometry.  I get the following `UniqueViolation` error, which I don't understand since I am trying to update that key-value.

`{'status': 'error', 'message': '(psycopg2.errors.UniqueViolation) duplicate key value violates unique constraint "deduplication_index"\nDETAIL:  Key (obj_id, instrument_id, origin, mjd, fluxerr, flux)=(DESI39627630114245131, 7, DESI, 59294.13, 0.0009132356328013504, 109.24501044004198) already exists.\n\n[SQL: INSERT INTO photometry (id, created_at, modified, ra, dec, mjd, flux, fluxerr, filter, ra_unc, dec_unc, original_user_data, altdata, upload_id, origin, obj_id, instrument_id, owner_id) VALUES (%(id)s, timezone(%(timezone_1)s, CURRENT_TIMESTAMP), timezone(%(timezone_1)s, CURRENT_TIMESTAMP), %(ra)s, %(dec)s, %(mjd)s, %(flux)s, %(fluxerr)s, %(filter)s, %(ra_unc)s, %(dec_unc)s, %(original_user_data)s, %(altdata)s, %(upload_id)s, %(origin)s, %(obj_id)s, %(instrument_id)s, %(owner_id)s)]\n[parameters: ({\'id\': 2795271, \'timezone_1\': \'UTC\', \'ra\': None, \'dec\': None, \'mjd\': 59294.13, \'flux\': 109.24501044004198, \'fluxerr\': 0.0009132356328013504, \'filter\': \'sdssg\', \'ra_unc\': None, \'dec_unc\': None, \'original_user_data\': \'{"magsys": "ab"}\', \'altdata\': \'{"subdir": "recent", "trunk": "coadd", "tile": "80647", "date": "20210320", "panel": "4"}\', \'upload_id\': \'9b5b1972-4f99-4fc2-a17b-379f56d6af57\', \'origin\': \'DESI\', \'obj_id\': \'DESI39627630114245131\', \'instrument_id\': 7, \'owner_id\': 384}, {\'id\': 2795272, \'timezone_1\': \'UTC\', \'ra\': None, \'dec\': None, \'mjd\': 59294.13, \'flux\': 109.24501044004198, \'fluxerr\': 0.0009132356328013504, \'filter\': \'sdssg\', \'ra_unc\': None, \'dec_unc\': None, \'original_user_data\': \'{"magsys": "ab"}\', \'altdata\': \'{"subdir": "recent", "trunk": "coadd", "tile": "80647", "date": "20210320", "panel": "4"}\', \'upload_id\': \'9b5b1972-4f99-4fc2-a17b-379f56d6af57\', \'origin\': \'DESI\', \'obj_id\': \'DESI39627630114245131\', \'instrument_id\': 7, \'owner_id\': 384})]\n(Background on this error at: http://sqlalche.me/e/gkpj)', 'data': {}, 'version': '0.9.dev0+git%aI.40fc611'}`

**Comment by acrellin** [comment]
Hi Alex, we need a minimum reproducible example in order to investigate
this. Can you please provide sample requests that will reproduce the
problem? Thank you!

**Comment by AlexGKim** [comment]
Here is an example in python, where I have removed my token.

```
In [14]: print(data)
{'instrument_id': 7, 'origin': 'DESI', 'zp': 0, 'mjd': '59294.13', 'fluxerr': 2.5152597927726703e-13, 'magsys': 'ab', 'group_ids': [76, 46], 'obj_id': 'DESI39627630114245131', 'filter': 'sdssg', 'flux': 3.0088574345044094e-08, 'altdata': {'tile': '80647', 'date': '20210320', 'panel': '4', 'subdir': 'recent', 'trunk': 'coadd'}}

In [15]: request = requests.request('PUT','http://desi2.lbl.gov:5000/api/photome
    ...: try',json=data,headers={'Authorization': 'token xxxx'})

In [16]: print (request.json())
{u'status': u'error', u'message': u'(psycopg2.errors.UniqueViolation) duplicate key value violates unique constraint "deduplication_index"\nDETAIL:  Key (obj_id, instrument_id, origin, mjd, fluxerr, flux)=(DESI39627630114245131, 7, DESI, 59294.13, 0.0009132356328013504, 109.24501044004198) already exists.\n\n[SQL: INSERT INTO photometry (id, created_at, modified, ra, dec, mjd, flux, fluxerr, filter, ra_unc, dec_unc, original_user_data, altdata, upload_id, origin, obj_id, instrument_id, owner_id) VALUES (%(id)s, timezone(%(timezone_1)s, CURRENT_TIMESTAMP), timezone(%(timezone_1)s, CURRENT_TIMESTAMP), %(ra)s, %(dec)s, %(mjd)s, %(flux)s, %(fluxerr)s, %(filter)s, %(ra_unc)s, %(dec_unc)s, %(original_user_data)s, %(altdata)s, %(upload_id)s, %(origin)s, %(obj_id)s, %(instrument_id)s, %(owner_id)s)]\n[parameters: ({\'id\': 2795281, \'timezone_1\': \'UTC\', \'ra\': None, \'dec\': None, \'mjd\': 59294.13, \'flux\': 109.24501044004198, \'fluxerr\': 0.0009132356328013504, \'filter\': \'sdssg\', \'ra_unc\': None, \'dec_unc\': None, \'original_user_data\': \'{"magsys": "ab"}\', \'altdata\': \'{"tile": "80647", "date": "20210320", "panel": "4", "subdir": "recent", "trunk": "coadd"}\', \'upload_id\': \'1d9069ed-0087-4c73-9a5d-9f1d7e6ec8c7\', \'origin\': \'DESI\', \'obj_id\': \'DESI39627630114245131\', \'instrument_id\': 7, \'owner_id\': 384}, {\'id\': 2795282, \'timezone_1\': \'UTC\', \'ra\': None, \'dec\': None, \'mjd\': 59294.13, \'flux\': 109.24501044004198, \'fluxerr\': 0.0009132356328013504, \'filter\': \'sdssg\', \'ra_unc\': None, \'dec_unc\': None, \'original_user_data\': \'{"magsys": "ab"}\', \'altdata\': \'{"tile": "80647", "date": "20210320", "panel": "4", "subdir": "recent", "trunk": "coadd"}\', \'upload_id\': \'1d9069ed-0087-4c73-9a5d-9f1d7e6ec8c7\', \'origin\': \'DESI\', \'obj_id\': \'DESI39627630114245131\', \'instrument_id\': 7, \'owner_id\': 384})]\n(Background on this error at: http://sqlalche.me/e/gkpj)', u'version': u'0.9.dev0+git%aI.40fc611', u'data': {}}
```

**Comment by AlexGKim** [comment]
My understanding is that PUT is supposed to update so I was not expecting the UniqueViolation.

**Comment by acrellin** [comment]
Please try omitting group_ids, which is not a valid parameter for that
endpoint (we'll update the error message to note this). But the unique
violation could still happen with a put request - we don't allow any two
photometry points to have the same `(obj_id, instrument_id, origin, mjd,
fluxerr, flux)` values - it's easy to create a collision with an existing
photometry point via a PUT request.

**Comment by acrellin** [comment]
@AlexGKim I misspoke earlier -- group_ids is a valid field. But the error message says it all: `Key (obj_id, instrument_id, origin, mjd, fluxerr, flux)=(DESI39627630114245131, 7, DESI, 59294.13, 0.0009132356328013504, 109.24501044004198) already exists`

No two points can have all of those values be exactly the same, otherwise we throw this error. Do you believe that is not the case, that this is being thrown erroneously?

**Comment by AlexGKim** [comment]
The values can be exactly the same but are not guaranteed to be.

**Comment by acrellin** [comment]
I'm not sure if I understand. We don't allow any two photometry points to have the same exact tuple of the above values. Does that answer the question, or do you think there isn't in fact a conflict between any two points and this error is arising when it shouldn't be?

**Comment by AlexGKim** [comment]
OK, I understand now.  It is unexpected behavior for me the user for an update to return an error if the same data is being "PUT".  Even though under the covers an update is not being done, to me I don't care as long as the data in my most recent PUT is saved.

**Comment by acrellin** [comment]
We actually are doing an update, it's just that if the above-mentioned tuple of the provided values is exactly the same as an existing photometry point (other than the one being updated), it'll be rejected, as we consider it a duplicate.

**Comment by acrellin** [comment]
@AlexGKim My apologies, after looking into this further, I have discovered that the issue is not what I had initially guessed, but is deeper and more real. We'll push a patch ASAP (this issue will be linked/closed by the PR that fixes it).

**Comment by AlexGKim** [comment]
Thanks @acrellin for the heads up.  I have not yet changed my code.  Will the new patch accept identical photometry updates, so I don't have to modify my code?

**Comment by acrellin** [comment]
@AlexGKim which fields differ in the PUT request payload from the originally provided POST payload?

**Comment by AlexGKim** [comment]
Sometimes no fields will differ.

We are reprocessing the photometry from the same observation multiple times.  However, the data extraction of that observation changes as upstream software improves.  The new extraction supersedes the old one.  When the observation photometry is processed, it may use an extraction for which photometry was already calculated and saved or it may use a fresh extraction whose new photometry must be saved.  What I would like is to PUT without having to construct an infrastructure to check if SkyPortal already contains an identical photometry.

**Comment by acrellin** [comment]
@AlexGKim Yep, makes sense. Thanks for the info. Will fix and update you here soon.

**Comment by AlexGKim** [comment]
It is not mission critical to fix on your end.  I have set up exception handling to catch this.  It is just that behavior is not intuitive to me so that documentation would be helpful.
